### PR TITLE
Bugfix for Linear, H1 and L2 functions on point meshes [bugfix-point-fespace]

### PR DESCRIPTION
**Pull request by kennyweiss**
For Linear and ``H1`` functions,  ``FiniteElementSpace`` was generating two dofs per point rather than one for ``LinearFECollection`` and (order 1) ``H1_FECollection``s  defined on a point mesh.

For ``L2`` functions,  ``L2_FECollection`` did not handle the 0-dimensional (point) case, causing the following code to segfault:

```cpp
// simple single element point mesh
int dim = 0, nv = 1, ne = 1, nb = 0, sdim = 2, order =1;
mfem::Mesh ptMesh(dim, nv, ne, nb, sdim);

ptMesh.AddVertex(mfem::Vertex(0,0)());
      
int idx[2] = {0};
ptMesh.AddElement(new mfem::Point(idx, 1));
 
// create a grid function using an L2 fec
mfem::L2_FECollection fec(0,dim);       //  <-- segfault here
mfem::FiniteElementSpace fes(&ptMesh, &fec);
mfem::GridFunction gf(&fes);
```

### CODE REVIEW

File: fem/fe_coll.cpp [hunk 1]
```diff
@@ -1742,7 +1742,11 @@ L2_FECollection::L2_FECollection(const int p, const int dim, const int type,
    }
    OtherDofOrd = NULL;
 
-   if (dim == 1)
+   if (dim == 0)
+   {
+      L2_Elements[Geometry::POINT] = new PointFiniteElement;
```
Comment (kennyweiss) [inline fem/fe_coll.cpp]: I don't really have enough context to know if this is sufficient to cover all the use cases of an L2 space.
It was sufficient to get rid of the segfault, and to assign one dof to my simple point mesh (see the PR description).

E.g.
* does it make sense to have a higher order L2 space on a point?
* are there other fields that need to be initialized here ( such as ``Tr_Elements``, which looks like it is only for lower dimensional faces)?

File: fem/fespace.cpp [hunk 2]
```diff
@@ -1059,7 +1059,8 @@ void FiniteElementSpace::Construct()
    bdofs[0] = 0;
    for (i = 0; i < mesh->GetNE(); i++)
    {
-      nbdofs += fec->DofForGeometry(mesh->GetElementBaseGeometry(i));
+      int geom = mesh->GetElementBaseGeometry(i);
+      nbdofs += geom == Geometry::POINT ? 0 : fec->DofForGeometry(geom);
```
Comment (tzanio) [inline fem/fespace.cpp]: Don't insist on it, but to me it is more clear to say:

```c++
if (geom != Geometry::POINT)
{
   nbdofs += fec->DofForGeometry(geom);
}
```

Comment (kennyweiss) [reply fem/fespace.cpp]: I agree, that is clearer. I'll change it.

**Comment by tzanio** [comment]
@rieben1, any objections to merging this?

**Comment by rieben1** [comment]
No, im fine with merging

**Comment by kennyweiss** [comment]
Thanks @v-dobrev.

I have a small update to add support for dumping point meshes in ``vtk`` format.

Should I add it to this PR, or create a separate one?

The changes are to add ``Geometry::POINT`` cases to the ``switch`` statements in:
* ``mesh::PrintVTK()`` 
* ``GeometryRefiner::Refine()``

Incidentally, your updates to ``FiniteElementSpace::GetElementDofs()`` were necessary to make this work.

**Comment by v-dobrev** [comment]
@kennyweiss, I'm OK with adding you updates here, if you want to.

File: fem/gridfunc.cpp [hunk 3]
```diff
@@ -2298,7 +2298,7 @@ void GridFunction::SaveVTK(std::ostream &out, const std::string &field_name,
          }
       }
    }
-   else if (vec_dim == mesh->Dimension())
+   else if ( (vec_dim == 2 || vec_dim == 3) && mesh->SpaceDimension() > 1)
```
Comment (kennyweiss) [inline fem/gridfunc.cpp]: This change allows 3D vector fields on surface meshes and point meshes.

File: mesh/mesh.cpp [hunk 4]
```diff
@@ -7485,23 +7489,27 @@ void Mesh::PrintVTK(std::ostream &out, int ref, int field_data)
...
 
-   Array<int> coloring;
-   srand((unsigned)time(0));
-   double a = double(rand()) / (double(RAND_MAX) + 1.);
-   int el0 = (int)floor(a * GetNE());
-   GetElementColoring(coloring, el0);
-   out << "SCALARS element_coloring int\n"
-       << "LOOKUP_TABLE default\n";
-   for (int i = 0; i < GetNE(); i++)
+   if (Dim > 1)
```
Comment (kennyweiss) [inline mesh/mesh.cpp]: There was no ElementToElement table for points or segments.

**Comment by kennyweiss** [comment]
@tzanio  @v-dobrev  @rieben1  -- I pushed some new changes to allow outputting VTK meshes for Point meshes.

**Comment by v-dobrev** [comment]
The new additions look good to me. Thanks, @kennyweiss !

**Pull request merged**